### ISSUE TITLE
Index API: make implicit ordering of responses explicit?

### ISSUE THREAD

**Issue by woodruffw** [pypi/warehouse#17258]
I'm opening this for discussion, based on a (non-breaking!) divergence between PyPI's observed behavior and the behavior specified in PEPs 503 and 691.

This comes from the context of Homebrew, which periodically pulls from the Index/JSON APIs as part of checking for releases. CC @samford for viz/more context.

TL;DR: right now, the response to `/simple/<name>/` is *ordered* by version:

https://github.com/pypi/warehouse/blob/e740cecc0931441d38426f163881d801d9ecefaf/warehouse/packaging/utils.py#L48-L65

As a result of this ordering, newer releases (by version) come last, and older releases come first.

Meanwhile, PEP 691 says the following about ordering:

> While the files key is an array, and thus is required to be in some kind of an order, neither [PEP 503](https://peps.python.org/pep-0503/) nor this PEP requires any specific ordering nor that the ordering is consistent from one request to the next. Mentally this is best thought of as a set, but both JSON and HTML lack the functionality to have sets.

In other words, this ordering is an implementation detail within PyPI. However, it's likely that consumers of the Index APIs are depending on this observed behavior (per [Hyrum's Law](https://www.hyrumslaw.com/)), and are using the current ordering as an optimization to more rapidly find the latest release(s). 

Given the above, I wonder if it makes sense for PyPI to make this implementation detail explicit. Some options:

* Document it in the Index API docs, as an implementation specific detail of PyPI's Index APIs
* Add some kind of additional field to the `meta` component of the JSON response, signaling that the index is ordering the fields, e.g.:

	```python
    "meta": {
      "api-version": "1.0",
	  "order": "version",
    },
    ```

	(This might be overkill though.)

OTOH, maybe this doesn't need to be documented or formalized. But I figured I'd open this up for consideration 🙂

**Comment by di** [comment]
> and are using the current ordering as an optimization to more rapidly find the latest release(s).

Just to be clear here, what the end user wants in this example is the _largest_ release (in terms of version) and not the _latest_ release (in terms of publication date), correct? Because they are not always the same 🙂.

**Comment by woodruffw** [comment]
> > and are using the current ordering as an optimization to more rapidly find the latest release(s).
> 
> Just to be clear here, what the end user wants in this example is the _largest_ release (in terms of version) and not the _latest_ release (in terms of publication date), correct? Because they are not always the same 🙂.

Correct! Sorry, bad phrasing on my part: I was using "latest" to mean "highest version" not "most recently published distribution." I'll update the issue body 🙂
